### PR TITLE
Implement destructive in-place rewrites for Cholesky and Solve Ops

### LINKED ISSUE #572: ENH: Add destructive rewrites for Cholesky `Op`
### Before

_No response_

### After

_No response_

### Context for the issue:

Currently, the Cholesky `Op` does not act in-place. This is a known problem since the Theano days (see comment [here](https://github.com/pymc-devs/pytensor/blob/9df55cc6c507ea5d78f88166664ab1325d1fd30d/pytensor/tensor/slinalg.py#L47-L49)). We should:

1. Add a rewrite that will set `destructive=True` to cholesky nodes, 
2. Add a destroy_map to Cholesky nodes
3. Register the rewrite somewhere after the specialization step (this is done by other destructive rewrites, e.g. for `GEMV`.

If anyone knows other linear algebra `Ops` that are not acting in-place, this issue could also collect those (`SolveTriangular`?)

### PR DESCRIPTION
**Pull request by ricardoV94**
## Description
Reviving accidentally closed #577 

All scipy.linalg functions offer an `overwrite_a` and/or `overwrite_b` argument that can enhance performance by re-using the input memory for the outputs.  This PR implements re-writes that will set these flags to True at compile time.

These rewrites are also a nice example for the in-place docs [here](https://pytensor.readthedocs.io/en/latest/extending/inplace.html), so I'll update them with an example as a later commit to this PR.

We added a general inplace functionality to Blockwise, that automatically makes the core_op inplace. This will translate to the core_op if the Blockwise is later removed due to being useless (no batch dims actually exist).

## Related Issue

- [x] Closes #572 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [x] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1028.org.readthedocs.build/en/1028/

### CODE REVIEW

File: pytensor/tensor/rewriting/blockwise.py [hunk 1]
```diff
@@ -225,3 +229,77 @@
...
+
+    if not inplace_core_op.destroy_map:
+        return None
+
+    # Check Op is not trying to inplace on non-candidate inputs
+    for destroyed_inputs in inplace_core_op.destroy_map.values():
+        for destroyed_input in destroyed_inputs:
+            if destroyed_input not in allowed_inplace_inputs:
+                raise ValueError(
+                    "Op destroy_map does not respect allowed_inplace_inputs"
```
Comment (jessegrabowski) [inline pytensor/tensor/rewriting/blockwise.py:282]: Can we provide some information about which Op is raising here? Something like `f"Op {op.__name__} destroy_map ...` ?

File: pytensor/tensor/blockwise.py [hunk 2]
```diff
@@ -79,6 +80,15 @@ def __init__(
         self.inputs_sig, self.outputs_sig = _parse_gufunc_signature(signature)
         self.gufunc_spec = gufunc_spec
         self._gufunc = None
+        if destroy_map is not None:
+            self.destroy_map = destroy_map
+        if self.destroy_map != core_op.destroy_map:
+            # Note: Should be fine for destroy_map of Blockwise to be more extensive than that of core_op
+            # But we are not using that anywhere yet, so this check is fine for now
+            raise ValueError(
+                "Blockwise destroy_map must be the same as that of the core_op"
```
Comment (jessegrabowski) [inline pytensor/tensor/blockwise.py:89]: Same here. A complaint I have in general is that our error messages are qutie cryptic.

File: pytensor/tensor/slinalg.py [hunk 3]
```diff
@@ -28,57 +28,68 @@
...
+    __props__ = ("lower", "check_finite", "on_error", "overwrite_a")
     gufunc_signature = "(m,m)->(m,m)"
 
-    def __init__(self, *, lower=True, check_finite=True, on_error="raise"):
+    def __init__(
+        self,
+        *,
+        lower: bool = True,
+        check_finite: bool = True,
+        on_error: Literal["raise"] | Literal["nan"] = "raise",
```
Comment (jessegrabowski) [inline pytensor/tensor/slinalg.py:41]: I don't think you have give one literal per possible value, you can do: `Literal["raise", "nan"]`

File: pytensor/tensor/slinalg.py [hunk 4]
```diff
@@ -131,11 +142,66 @@ def conjugate_solve_triangular(outer, inner):
...
+        import numpy as np
+
+        x = pt.tensor('x', shape=(5, 5), dtype='float64')
+        L = pt.linalg.cholesky(x)
+
+        f = pytensor.function([x], L)
+        x_value = np.random.normal(size=(5, 5))
+        x_value = x_value @ x_value.T # Ensures x is positive definite
+        L_value = f(x_value)
+        assert np.allclose(L_value @ L_value.T, x_value)
```
Comment (jessegrabowski) [inline pytensor/tensor/slinalg.py:200]: Replace with a `.. testoutput::` block?

File: pytensor/tensor/slinalg.py [hunk 5]
```diff
@@ -162,9 +232,25 @@ def __init__(
...
         else:
             self.gufunc_signature = "(m,m),(m,n)->(m,n)"
+        self.overwrite_a = overwrite_a
+        self.overwrite_b = overwrite_b
+        destroy_map = {}
+        if self.overwrite_a and self.overwrite_b:
+            # An output destroying two inputs is not yet supported
+            # destroy_map[0] = [0, 1]
+            raise NotImplementedError(
+                "It's not yet possible to overwrite_a and overwrite_b symultaneously"
```
Comment (jessegrabowski) [inline pytensor/tensor/slinalg.py:242]: typo: simultaneously

**Review by jessegrabowski — APPROVED** [review]
Very cool, glad to see this over the finish line (with a bunch of extended functionality!) 

I left mostly nitpicks. This could be merged as-is without any changes IMO. 

My biggest criticism of the approach, if I had to offer one, would be the use of a list of integers for `allowed_inplace_inputs`, which you seem to be aware of yoruself because of the extended comment in `slinalg.py:L521-523`. I don't have a better alternative (what came immediately to mind is to always pass a list of bool of length `len(self.inputs)`, and check the appropriate positions for True. but it's no better). I guess it's frustrating that we sometimes encode informaton about codes with dicts (`destroy_map`) and other times with lists (`allowed_inplace_inputs`). It's somewhat difficult as a dev to know what to expect.

File: pytensor/tensor/slinalg.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:200]: To show it outputs True from `np.allclose`? I put an assert because the output was not interesting here

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:200]: Does doctest still run this without the testoutput? That was the assumption behind this comment, ignore if wrong.

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:200]: Yes it does

Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:200]: https://github.com/pymc-devs/pytensor/actions/runs/11277835281/job/31364961354?pr=1028#step:6:58

**Pull request merged**